x/qgb/keeper/query_valset.go

```diff
@@ -41,3 +42,22 @@ func (k Keeper) ValsetRequestByNonce(
 
 	return nil, sdkerrors.Wrap(sdkerrors.ErrNotFound, "valset request nonce not found")
 }
+
+// LastValsetBeforeHeight queries the last valset request before height
+func (k Keeper) LastValsetBeforeHeight(
+	c context.Context,
+	req *types.QueryLastValsetBeforeHeightRequest) (*types.QueryLastValsetBeforeHeightResponse, error) {
+	valReq := k.GetValsets(sdk.UnwrapSDKContext(c))
+	for _, valset := range valReq {
```

Comment:
This seems rather inefficient, having to iterate over all valsets ever.

Reply:
Hmm I guess that's true. As Satoshi would say,

> With computer systems typically selling with 2GB of RAM as of 2008, and Moore's Law predicting current growth of
1.2GB per year, storage should not be a problem even if the [validator sets] must be kept in
memory.

Reply:
We still have a fair amount of work to do for optimizing these things. As we've previously discussed, we could change the way we index so that we make better use of maps instead of relying heavily on iterators, but this will require a significant refactor and departure from the gravity module's code. It won't hurt to keep track of instances where we can improve, so I'll link #369